File: mountainview/src/views/mvpanelwidget.cpp
/******************************************************
** See the accompanying README and LICENSE files
** Author(s): Jeremy Magland
** Created: 7/29/2016
*******************************************************/

#include "mvpanelwidget.h"
#include "paintlayerstack.h"
#include <QRectF>
#include <QDebug>

struct MVPanelWidgetPanel {
    int row = 0;
    int col = 0;
    PaintLayer* layer = 0;
    QRectF geom = QRectF(0, 0, 0, 0);
};

class MVPanelWidgetPrivate {
public:
    MVPanelWidget* q;

    QList<MVPanelWidgetPanel> m_panels;
    int m_row_margin = 3;
    int m_row_spacing = 3;
    int m_col_margin = 3;
    int m_col_spacing = 3;
    double m_max_panel_width = 600;
    QRectF m_viewport_geom = QRectF(0, 0, 1, 1);
    int m_current_panel_index = -1;

    void correct_viewport_geom();
    void zoom(double factor);
    void update_panel_geometries();
    int panel_index_at_pt(const QPointF& pt);
};

MVPanelWidget::MVPanelWidget()
{
    d = new MVPanelWidgetPrivate;
    d->q = this;
}

MVPanelWidget::~MVPanelWidget()
{
    delete d;
}

void MVPanelWidget::clearPanels(bool delete_layers)
{
    if (delete_layers) {
        foreach (MVPanelWidgetPanel panel, d->m_panels) {
            delete panel.layer;
        }
    }
    d->m_panels.clear();
    update();
}

void MVPanelWidget::addPanel(int row, int col, PaintLayer* layer)
{
    MVPanelWidgetPanel panel;
    panel.row = row;
    panel.col = col;
    panel.layer = layer;
    d->m_panels << panel;
    update();
}

int MVPanelWidget::rowCount() const
{
    int ret = 0;
    for (int i = 0; i < d->m_panels.count(); i++) {
        ret = qMax(ret, d->m_panels[i].row + 1);
    }
    return ret;
}

int MVPanelWidget::columnCount() const
{
    int ret = 0;
    for (int i = 0; i < d->m_panels.count(); i++) {
        ret = qMax(ret, d->m_panels[i].col + 1);
    }
    return ret;
}

void MVPanelWidget::setSpacing(int row_spacing, int col_spacing)
{
    d->m_row_spacing = row_spacing;
    d->m_col_spacing = col_spacing;
    update();
}

void MVPanelWidget::setMargins(int row_margin, int col_margin)
{
    d->m_row_margin = row_margin;
    d->m_col_margin = col_margin;
    update();
}

void MVPanelWidget::setViewportGeometry(QRectF geom)
{
    d->m_viewport_geom = geom;
    update();
}

void MVPanelWidget::paintEvent(QPaintEvent* evt)
{
    Q_UNUSED(evt)
    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing);
    d->correct_viewport_geom();
    d->update_panel_geometries();

    for (int i = 0; i < d->m_panels.count(); i++) {
        QRectF geom = d->m_panels[i].geom;
        d->m_panels[i].layer->setWindowSize(geom.size().toSize());
        double x1 = geom.left();
        double y1 = geom.top();
        painter.translate(QPointF(x1, y1));
        d->m_panels[i].layer->paint(&painter);
        painter.translate(QPointF(-x1, -y1));
    }
}

void MVPanelWidget::wheelEvent(QWheelEvent* evt)
{
    int delta = evt->delta();
    double factor = 1;
    if (delta > 0)
        factor = 1.2;
    else
        factor = 1 / 1.2;
    d->zoom(factor);
}

void MVPanelWidget::mousePressEvent(QMouseEvent* evt)
{
    qDebug() << __FUNCTION__ << __FILE__ << __LINE__;
    int index = d->panel_index_at_pt(evt->pos());
    qDebug() << __FUNCTION__ << __FILE__ << __LINE__ << ":::::::::::::::" << index;
    if (index >= 0) {
        d->m_current_panel_index = index;
        update();
    }
    QWidget::mousePressEvent(evt);
}

void MVPanelWidgetPrivate::correct_viewport_geom()
{
    QRectF G = m_viewport_geom;
    if (q->width() * G.width() / m_panels.count() > m_max_panel_width) {
        G.setWidth(m_max_panel_width * m_panels.count() / q->width());
    }
    if (G.width() < 1) {
        G.setLeft(0);
        G.setWidth(1);
    }
    if (G.left() + G.width() < 1)
        G.setLeft(1 - G.width());
    if (G.left() > 0)
        G.setLeft(0);
    if (G.height() < 1) {
        G.setTop(0);
        G.setHeight(1);
    }
    if (G.top() + G.height() < 1)
        G.setTop(1 - G.height());
    if (G.top() > 0)
        G.setTop(0);
    m_viewport_geom = G;
}

void MVPanelWidgetPrivate::zoom(double factor)
{
    QRectF current_panel_geom = QRectF(0, 0, 1, 1);
    if ((m_current_panel_index >= 0) && (m_current_panel_index < m_panels.count())) {
        current_panel_geom = m_panels[m_current_panel_index].geom;
    }
    if (q->rowCount() > 1) {
        m_viewport_geom.setHeight(m_viewport_geom.height() * factor);
    }
    if (q->columnCount() > 1) {
        m_viewport_geom.setWidth(m_viewport_geom.width() * factor);
    }
    correct_viewport_geom();
    update_panel_geometries();
    QRectF new_current_panel_geom = QRectF(0, 0, 1, 1);
    if ((m_current_panel_index >= 0) && (m_current_panel_index < m_panels.count())) {
        new_current_panel_geom = m_panels[m_current_panel_index].geom;
    }
    double dx = new_current_panel_geom.center().x() - current_panel_geom.center().x();
    double dy = new_current_panel_geom.center().y() - current_panel_geom.center().y();
    if ((dx) || (dy)) {
        m_viewport_geom.setLeft(m_viewport_geom.left() - dx / q->width());
        m_viewport_geom.setTop(m_viewport_geom.top() - dy / q->height());
    }
    correct_viewport_geom();
    update_panel_geometries();
    q->update();
}

void MVPanelWidgetPrivate::update_panel_geometries()
{
    int num_rows = q->rowCount();
    int num_cols = q->columnCount();
    if (!num_rows)
        return;
    if (!num_cols)
        return;
    double viewport_W = q->width();
    double viewport_H = q->height();
    QRectF VG = m_viewport_geom;
    QRectF vgeom = QRectF(VG.left() * viewport_W, VG.top() * viewport_H, VG.width() * viewport_W, VG.height() * viewport_H);
    double W0 = (vgeom.width() - (num_cols - 1) * m_col_spacing - m_col_margin * 2) / num_cols;
    double H0 = (vgeom.height() - (num_rows - 1) * m_row_spacing - m_row_margin * 2) / num_rows;
    for (int i = 0; i < m_panels.count(); i++) {
        double x1 = vgeom.x() + m_col_margin + m_panels[i].col * (W0 + m_col_spacing);
        double y1 = vgeom.y() + m_row_margin + m_panels[i].row * (H0 + m_row_spacing);
        m_panels[i].geom = QRectF(x1, y1, W0, H0);
    }
}

int MVPanelWidgetPrivate::panel_index_at_pt(const QPointF& pt)
{
    for (int i = 0; i < m_panels.count(); i++) {
        if (m_panels[i].geom.contains(pt))
            return i;
    }
    return -1;
}

File: WIP/web/jsq/mountainviewweb/core/mvpanelwidget.js
function MVPanelWidget(O) {
	if (!O) O=this;
	JSQWidget(O);

	O.clearPanels=function() {clearPanels();};
	O.addPanel=function(row,col,W) {addPanel(row,col,W);};
	O.rowCount=function() {return rowCount();};
	O.columnCount=function() {return columnCount();};
	O.setSpacing=function(row_spacing,col_spacing) {m_row_spacing=row_spacing; m_col_spacing=m_col_spacing; update_layout();};
	O.setMargins=function(row_margin,col_margin) {m_row_margin=row_margin; m_col_margin=col_margin;};

	O.onPanelClicked=function(handler) {JSQ.connect(O,'panel_clicked',O,function(sender,args) {handler(args.ind,args.modifiers);});};

	O.onMousePress(mousePress);
	O.onMouseRelease(mouseRelease);
	O.onMouseMove(mouseMove);
	O.onMouseEnter(mouseEnter);
	O.onMouseLeave(mouseLeave);
	O.onWheel(wheel);

	var m_panels=[];
	var m_row_margin=3;
	var m_row_spacing=3;
	var m_col_margin=3;
	var m_col_spacing=3;
	var m_viewport_geom=[0,0,1,1];
	var m_current_panel_index=-1;

	JSQ.connect(O,'sizeChanged',O,update_layout);
	function update_layout() {
		correct_viewport_geom();
		var num_rows=O.rowCount();
		var num_cols=O.columnCount();
		if (!num_rows) return;
		if (!num_cols) return;
		var VP_W=O.size()[0];
		var VP_H=O.size()[1];
		var VG=m_viewport_geom;
		vgeom=[VG[0]*VP_W,VG[1]*VP_H,VG[2]*VP_W,VG[3]*VP_H];
		var W0=(vgeom[2]-(num_cols-1)*m_col_spacing-m_col_margin*2)/num_cols;
		var H0=(vgeom[3]-(num_rows-1)*m_row_spacing-m_row_margin*2)/num_rows;
		for (var i=0; i<m_panels.length; i++) {
			var x1=vgeom[0] + m_col_margin + m_panels[i].col*(W0+m_col_spacing);
			var y1=vgeom[1] + m_row_margin + m_panels[i].row*(H0+m_row_spacing);
			
			m_panels[i].W.setSize([W0,H0]);
			m_panels[i].W.setPosition([x1,y1]);
			m_panels[i].pixel_geom=[x1,y1,W0,H0];
		}
	}
	function correct_viewport_geom() {
		if (m_viewport_geom[2]<1) {
			m_viewport_geom[0]=0;
			m_viewport_geom[2]=1;
		}
		if (m_viewport_geom[0]+m_viewport_geom[2]<1) {
			m_viewport_geom[0]=1-m_viewport_geom[2];
		}
		if (m_viewport_geom[0]>0) {
			m_viewport_geom[0]=0;
		}
		if (m_viewport_geom[3]<1) {
			m_viewport_geom[1]=0;
			m_viewport_geom[3]=1;
		}
		if (m_viewport_geom[1]+m_viewport_geom[3]<1) {
			m_viewport_geom[1]=1-m_viewport_geom[3];
		}
		if (m_viewport_geom[1]>0) {
			m_viewport_geom[1]=0;
		}
	}
	function clearPanels() {
		for (var i=0; i<m_panels.length; i++) {
			m_panels[i].W.destroy();
		}
		m_panels=[];
		update_layout();
	}
	function addPanel(row,col,W) {
		var panel={row:row,col:col,W:W,pixel_geom:[0,0,0,0]};
		m_panels.push(panel);
		O.div().append(W.div());
		W.div().addClass("mvpanelwidget_panel");
		/// TODO, use a queued signal so we don't update layout after every add or resize
		update_layout();
	}
	function rowCount() {
		var ret=0;
		for (var i=0; i<m_panels.length; i++) {
			if (m_panels[i].row+1>ret) ret=m_panels[i].row+1;
		}
		return ret;
	}
	function columnCount() {
		var ret=0;
		for (var i=0; i<m_panels.length; i++) {
			if (m_panels[i].col+1>ret) ret=m_panels[i].col+1;
		}
		return ret;
	}
	function zoom(factor) {
		var current_panel_geom=[0,0,1,1];
		if ((m_current_panel_index>=0)&&(m_current_panel_index<m_panels.length)) {
			current_panel_geom=JSQ.clone(m_panels[m_current_panel_index].pixel_geom);
		}

		if (rowCount()>1) {
			m_viewport_geom[3]=m_viewport_geom[3]/factor;
		}
		if (columnCount()>1) {
			m_viewport_geom[2]=m_viewport_geom[2]/factor;
		}
		update_layout();

		var new_current_panel_geom=[0,0,1,1];
		if ((m_current_panel_index>=0)&&(m_current_panel_index<m_panels.length)) {
			new_current_panel_geom=JSQ.clone(m_panels[m_current_panel_index].pixel_geom);
		}
		var dx=new_current_panel_geom[0]+new_current_panel_geom[2]/2-current_panel_geom[0]-current_panel_geom[2]/2;
		var dy=new_current_panel_geom[1]+new_current_panel_geom[3]/2-current_panel_geom[1]-current_panel_geom[3]/2;
		if ((dx)||(dy)) {
			m_viewport_geom[0]-=dx/O.size()[0];
			m_viewport_geom[1]-=dy/O.size()[1];
			update_layout();
		}
		correct_viewport_geom();
	}
	function set_current_panel_index(ind) {
		if (m_current_panel_index==ind) return;
		m_current_panel_index=ind;
		for (var ii=0; ii<m_panels.length; ii++) {
			if (ii==ind) {
				m_panels[ii].W.div().addClass("mvpanelwidget_current_panel");
			}
			else {
				m_panels[ii].W.div().removeClass("mvpanelwidget_current_panel");	
			}
		}

	}
	function rect_contains(rect,pt) {
		if ((pt[0]<rect[0])||(pt[0]>=rect[0]+rect[2])) return false;
		if ((pt[1]<rect[1])||(pt[1]>=rect[1]+rect[3])) return false;
		return true;
	}
	function panel_index_at(pos) {
		for (var i=0; i<m_panels.length; i++) {
			if (rect_contains(m_panels[i].pixel_geom,pos)) return i;
		}
		return -1;
	}

	var press_anchor=[-1,-1];
	var press_anchor_viewport_geom=[0,0,1,1];
	var is_dragging=false;
	function mousePress(evt) {
		press_anchor=JSQ.clone(evt.pos);
		press_anchor_viewport_geom=JSQ.clone(m_viewport_geom);
		is_dragging=false;
	}
	function mouseRelease(evt) {
		press_anchor=[-1,-1];
		if (!is_dragging) {
			var ind=panel_index_at(evt.pos);
			set_current_panel_index(ind);
			O.emit('panel_clicked',{ind:ind,modifiers:evt.modifiers});
		}
		is_dragging=false;
	}
	function mouseMove(evt) {
		if (press_anchor[0]>=0) {
			var dx=evt.pos[0]-press_anchor[0];
			var dy=evt.pos[1]-press_anchor[1];

			if ((Math.abs(dx)>=4)||(Math.abs(dy)>=4)) {
				is_dragging=true;
			}
			if (is_dragging) {
				if (columnCount()>1) {
					m_viewport_geom[0]=press_anchor_viewport_geom[0]+dx/O.size()[0];
				}
				if (rowCount()>1) {
					m_viewport_geom[1]=press_anchor_viewport_geom[1]+dy/O.size()[1];
				}
				update_layout();
			}
		}
	}
	function mouseEnter(evt) {

	}
	function mouseLeave(evt) {
		press_anchor=[-1,-1];
	}
	function wheel(evt) {
		if (evt.delta>0) {
			zoom(1/1.2);
		}
		else if (evt.delta<0) {
			zoom(1.2);
		}
	}

}